earth_enterprise/src/common/khxml/khxml.cpp

```diff
@@ -101,66 +126,191 @@ void setDefaultValues()
+// cede control  back to the control of write operations.
+//
+// TODO: generalize and move read-write lock to khThread
+
+static pthread_rwlock_t purge_lock = PTHREAD_RWLOCK_INITIALIZER;
+
+class XmlWriteLock
+{
+private:
+    pthread_rwlock_t& lock;
+public:
+    XmlWriteLock(pthread_rwlock_t& _lock) : lock(_lock)
+    {
+        pthread_rwlock_rdlock(&lock);
+    }
+    
+    ~XmlWriteLock()
+    {
+        pthread_rwlock_unlock(&lock);
+    }
+};
+
+class PurgeLock
+{
+private:
+    pthread_rwlock_t& lock;
+public:
+    PurgeLock(pthread_rwlock_t& _lock) : lock(_lock)
+    {
+        pthread_rwlock_wrlock(&lock);
+    }
+
+    ~PurgeLock()
+    {
+        pthread_rwlock_unlock(&lock);
+    }
+};
+
+void ReInitializeXerces()
+{
+    khLockGuard guard(reinitLock);
+    try
+    {
+        XMLPlatformUtils::Terminate();
+    }
+    catch(...)
+    {
+        notify(NFY_WARN, "Failure to terminate in ReInitializeXerces()");
+    }
+    initXercesValues();
+}
+
+
 // This is used only in the following function
 class UsingXMLGuard
 {
   friend void InitializeXMLLibrary() throw();
 
   UsingXMLGuard(void) throw() {
-    setDefaultValues();
-    std::string fn(XMLConfigFile);
-    try {
-      for (const auto& a : options)
-      {
-          config_parser.addOption(a);
-      }
-      config_parser.parse(fn);
-      config_parser.validateIntegerValues();
-      for (const auto& a : config_parser)
-      {
-          // will only set the values that are present, otherwise will stay at defaults
-          if (a.first == INIT_HEAP_SIZE)
-              initialDOMHeapAllocSize = std::stol(a.second);
-          else if (a.first == MAX_HEAP_SIZE)
-              maxDOMHeapAllocSize = std::stol(a.second);
-          else if (a.first == BLOCK_SIZE)
-              maxDOMSubAllocationSize = std::stol(a.second);
-      }
-    } catch (const khConfigFileParserException& e) {
-      notify(NFY_DEBUG, "%s , using default xerces init values", e.what());
-    }
-    
-    try {
-      try {
-         validateXMLParameters();
-      } catch (const XmlParamsException& e) {
-         notify(NFY_DEBUG, "%s, using default xerces init values", e.what());
-         setDefaultValues();
-      }
-      XMLPlatformUtils::Initialize(initialDOMHeapAllocSize,
-                                   maxDOMHeapAllocSize,
-                                   maxDOMSubAllocationSize);
-      notify(NFY_DEBUG, "XML initialization values: %s=%zu %s=%zu %s=%zu",
-             "initialDOMHeapAllocSize", initialDOMHeapAllocSize,
-             "maxDOMHeapAllocSize", maxDOMHeapAllocSize,
-             "maxDOMSubAllocationSize", maxDOMSubAllocationSize);
-    } catch(const XMLException& toCatch) {
-      notify(NFY_FATAL, "Unable to initialize Xerces: %s",
-             FromXMLStr(toCatch.getMessage()).c_str());
-    }
+    initXercesValues();
   }
 
   ~UsingXMLGuard(void) throw() {
-    try {
-      XMLPlatformUtils::Terminate();
-    } catch (...) {
+    if (!terminateCache)
```

Comment:
Minor: I don't know that this check is necessary. We can call terminate when the program shuts down regardless of whether we call terminate at other times as well.